File: 第14课 课件+四种语言模板+代码/【并查集模板】lc547 省份数量.cpp
class Solution {
public:
    int findCircleNum(vector<vector<int>>& isConnected) {
        // 建立并查集
        for (int i = 0; i < isConnected.size(); i++)
            fa.push_back(i);
        // 每条边代表一次合并
        for (int i = 0; i < isConnected.size(); i++)
            for (int j = i + 1; j < isConnected.size(); j++)
                if (isConnected[i][j] == 1) { 
                    unionSet(i, j);
                }
        // 有几棵树？（有几个根）
        int ans = 0;
        for (int i = 0; i < isConnected.size(); i++)
            if (find(i) == i) ans++;
        return ans;
    }

private:
    void unionSet(int i, int j) {
        int x = find(i);
        int y = find(j);
        if (x != y) fa[x] = y;
    }

    int find(int x) {
        if (x == fa[x]) return x;
        return fa[x] = find(fa[x]);
    }

    vector<int> fa;
};

File: 第14课 课件+四种语言模板+代码/【并查集模板】lc547 省份数量.java
class Solution {
    public int findCircleNum(int[][] isConnected) {
        DisjointSet s = new DisjointSet(isConnected.length);
        for (int i = 0; i < isConnected.length; i++)
            for (int j = 0; j < isConnected.length; j++)
                if (isConnected[i][j] == 1) s.unionSet(i, j);
        int ans = 0;
        for (int i = 0; i < isConnected.length; i++)
            if (s.find(i) == i) ans++;
        return ans;
    }

class DisjointSet {
    public DisjointSet(int n) {
        fa = new int[n];
        for (int i = 0; i < n; i++) fa[i] = i;
    }

    public int find(int x) {
        if (x == fa[x]) return x;
        return fa[x] = find(fa[x]);
    }

    public void unionSet(int x, int y) {
        x = find(x);
        y = find(y);
        if (x != y) fa[x] = y;
    }

    int[] fa;
};
}